ambry-store/src/main/java/com.github.ambry.store/ScanResults.java

```diff
@@ -0,0 +1,137 @@
+/**
+ * Copyright 2017 LinkedIn Corp. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ */
+
+package com.github.ambry.store;
+
+import com.github.ambry.utils.Pair;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.NavigableMap;
+import java.util.TreeMap;
+
+
+/**
+ * Hold the data structures needed by {@link BlobStoreStats} to serve requests. The class also holds helper methods
+ * used to modify and access the stored data structures.
+ */
+class ScanResults {
+  private final NavigableMap<Long, Map<String, Map<String, Long>>> containerBuckets = new TreeMap<>();
+  private final NavigableMap<Long, NavigableMap<String, Long>> logSegmentBuckets = new TreeMap<>();
+  final long containerForecastStartTimeMs;
+  final long containerLastBucketTimeMs;
+  final long containerForecastEndTimeMs;
+  final long logSegmentForecastStartTimeMs;
+  final long logSegmentLastBucketTimeMs;
+  final long logSegmentForecastEndTimeMs;
+  Offset lastScannedOffset = null;
+
+  /**
+   * Create the bucket data structures in advance based on the given scanStartTime and segmentScanTimeOffset.
+   */
+  ScanResults(long startTimeInMs, long logSegmentForecastOffsetMs, int bucketCount, long bucketSpanInMs) {
+    long containerBucketTimeMs = startTimeInMs;
+    long logSegmentBucketTimeMs = startTimeInMs - logSegmentForecastOffsetMs;
+    for (int i = 0; i < bucketCount; i++) {
+      containerBuckets.put(containerBucketTimeMs, new HashMap<String, Map<String, Long>>());
+      logSegmentBuckets.put(logSegmentBucketTimeMs, new TreeMap<String, Long>());
+      containerBucketTimeMs += bucketSpanInMs;
+      logSegmentBucketTimeMs += bucketSpanInMs;
+    }
+    containerForecastStartTimeMs = containerBuckets.firstKey();
+    containerLastBucketTimeMs = containerBuckets.lastKey();
+    containerForecastEndTimeMs = containerLastBucketTimeMs + bucketSpanInMs;
+    logSegmentForecastStartTimeMs = logSegmentBuckets.firstKey();
+    logSegmentLastBucketTimeMs = logSegmentBuckets.lastKey();
+    logSegmentForecastEndTimeMs = logSegmentLastBucketTimeMs + bucketSpanInMs;
+  }
+
+  Long getContainerBucketKey(long referenceTimeInMs) {
+    return containerBuckets.higherKey(referenceTimeInMs);
```

Comment:
hope you know the diff between ceilingKey() and higherKey() and you are using higherKey intentionally